src/chainstate/burn/db/burndb.rs

```diff
@@ -0,0 +1,1767 @@
+        tx.commit()
+            .map_err(db_error::SqliteError);
+
+        Ok(())
+    }
+
+    /// Open the burn database at the given path.  Open read-only or read/write.
+    /// If opened for read/write and it doesn't exist, instantiate it.
+    pub fn connect(path: &String, first_block_height: u64, first_burn_hash: &BurnchainHeaderHash, readwrite: bool) -> Result<BurnDB<A, K>, db_error> {
+        let mut create_flag = false;
+        let open_flags =
+            if fs::metadata(path).is_err() {
+                // need to create 
+                if readwrite {
+                    create_flag = true;
+                    OpenFlags::SQLITE_OPEN_READ_WRITE | OpenFlags::SQLITE_OPEN_CREATE
+                }
+                else {
+                    return Err(db_error::NoDBError);
+                }
+            }
+            else {
+                // can just open 
+                if readwrite {
+                    OpenFlags::SQLITE_OPEN_READ_WRITE
+                }
+                else {
+                    OpenFlags::SQLITE_OPEN_READ_ONLY
+                }
+            };
+
+        let conn = Connection::open_with_flags(path, open_flags)
+            .map_err(|e| db_error::SqliteError(e))?;
+
+        let mut db = BurnDB {
+            conn: conn,
+            readwrite: readwrite,
+            tx_active: false,
+            first_block_height: first_block_height,
+            first_burn_header_hash: first_burn_hash.clone(),
+            
+            _phantom_a: PhantomData,
+            _phantom_k: PhantomData
+        };
+
+        if create_flag {
+            // instantiate!
+            db.instantiate()?;
+        }
+        Ok(db)
+    }
+
+    /// Open a burn database in memory (used for testing)
+    pub fn connect_memory(first_block_height: u64, first_burn_hash: &BurnchainHeaderHash) -> Result<BurnDB<A, K>, db_error> {
+        let conn = Connection::open_in_memory()
+            .map_err(|e| db_error::SqliteError(e))?;
+
+        let mut db = BurnDB {
+            conn: conn,
+            readwrite: true,
+            tx_active: false,
+            first_block_height: first_block_height,
+            first_burn_header_hash: first_burn_hash.clone(),
+            
+            _phantom_a: PhantomData,
+            _phantom_k: PhantomData
+        };
+
+        db.instantiate()?;
+        Ok(db)
+    }
+
+    pub fn conn(&self) -> &Connection {
+        &self.conn
+    }
+
+    /// Get the previous consensus hashes that must be hashed to find
+    /// the *next* consensus hash at a particular block.
+    /// The resulting list will include the consensus hash at block_height.
+    fn get_prev_consensus_hashes(conn: &Connection, block_height: u64, first_block_height: u64) -> Result<Vec<ConsensusHash>, db_error> {
+        let mut i = 0;
+        let mut prev_chs = vec![];
+        while block_height - (((1 as u64) << i) - 1) >= first_block_height {
```

Comment:
Should the first block always be included in the list of consensus hashes?

Reply:
The caller is expected to pass a value for `block_height` for which a consensus hash already exists (i.e. there will be a row in the `snapshots` table for it).

Reply:
I meant `first_block_height` -- the goal of this is forming a merkle skip-list all the way back to the genesis block, right? In that case, shouldn't it always include the genesis block?

Reply:
The start of the skip list is the _last_ block, not the first block.  It does not always include the genesis block; it always includes the last-inserted block.

Reply:
Ah, I see now.